src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotIgnoreReturnValue.cs

```diff
@@ -0,0 +1,108 @@
+    {
+        internal const string CA2022RuleId = "CA2022";
+        private static readonly LocalizableString s_title = CreateLocalizableResourceString(nameof(DoNotIgnoreReturnValueTitle));
+        private static readonly LocalizableString s_description = CreateLocalizableResourceString(nameof(DoNotIgnoreReturnValueDescription));
+
+        internal static readonly DiagnosticDescriptor DoNotIgnoreReturnValueRule = DiagnosticDescriptorHelper.Create(
+            CA2022RuleId,
+            s_title,
+            CreateLocalizableResourceString(nameof(DoNotIgnoreReturnValueMessage)),
+            DiagnosticCategory.Reliability,
+            RuleLevel.BuildWarning,
+            s_description,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        internal static readonly DiagnosticDescriptor DoNotIgnoreReturnValueRuleWithMessage = DiagnosticDescriptorHelper.Create(
+            CA2022RuleId,
+            s_title,
+            CreateLocalizableResourceString(nameof(DoNotIgnoreReturnValueMessageCustom)),
+            DiagnosticCategory.Reliability,
+            RuleLevel.BuildWarning,
+            s_description,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(DoNotIgnoreReturnValueRule, DoNotIgnoreReturnValueRuleWithMessage);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemDiagnosticsCodeAnalysisDoNotIgnoreAttribute, out var doNotIgnoreAttribute))
+                {
+                    return;
+                }
+
+                context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksTask, out var taskType);
+
+                context.RegisterOperationAction(context =>
+                {
+                    var callSite = context.Operation;
+                    var method = ((IInvocationOperation)callSite).TargetMethod;
+
+                    // It would be an authoring error, but ensure the method returns a value
+                    if (method.ReturnsVoid)
+                    {
+                        return;
+                    }
+
+                    // If the method is async and awaited, then we'll check the await result for consumption
+                    if (method.IsAsync && callSite.Parent?.Kind is OperationKind.Await)
+                    {
+                        // It would be an authoring error, but ensure the async method returns a value (and not simply a Task)
+                        if (taskType?.Equals(method.ReturnType, SymbolEqualityComparer.Default) ?? false)
```

Comment:
Instead of this, I'd get the type of the `IAwaitOperation` and check if it's void.

Reply:
Also this condition checks `Task` only, so a corresponding `ValueTask` scenario may be incorrectly producing a diagnostic. Consider a ValueTask test similar to `AnnotatedAsyncMethod_WithoutReturnValue_NoDiagnostic`.

Reply:
> Instead of this, I'd get the type of the `IAwaitOperation` and check if it's void.

Very cool idea; thanks!

Reply:
I've changed to the following:

```c#
// If the method is async and awaited, then we'll check the await result for consumption
if (method.IsAsync && callSite.Parent?.Kind is OperationKind.Await)
{
  // It would be an authoring error, but ensure the async method returns a value
  if (((IAwaitOperation)callSite.Parent).Type.SpecialType is SpecialType.System_Void)
  {
    return;
  }

  callSite = callSite.Parent;
}
```

Given the C# test, this is behaving as I expect, recognizing that the async method will result in an `int`:

```c#
class C
{
  [return: DoNotIgnore]
  async Task<int> AnnotatedAsyncMethod() => 1;

  async void M()
  {
    await AnnotatedAsyncMethod(); // DIAGNOSTIC PRODUCED
                                  // IAwaitOperation.Type == int
                                  // IAwaitOperation.Operation.Type == Task<int>.
  }
}
```

But the corresponding VB test is not behaving that way. Instead, it is indicating that the operation type is `void`:

```vb
Public Class C
  Async Function AnnotatedAsyncMethod() As <DoNotIgnore> Task(Of Integer)
    Return 1
  End Function
                                
  Async Sub M()
    Await AnnotatedAsyncMethod() ' NO DIAGNOSTIC.
                                 ' IAwaitOperation.Type == void
                                 ' IAwaitOperation.Operation.Type = Task<int>
  End Sub
End Class
```

I would expect the VB scenario to also give `IAwaitOperation.Type` of `int`. Does anything stand out here, or is it possible this is a CodeAnalysis bug?

Reply:
@jeffhandley I have no idea. I think maybe @333fred can help?

Reply:
Relevant compiler code

https://github.com/dotnet/roslyn/blob/37bfc47fa1e81ef7807f0cf8d55ce0f184db6303/src/Compilers/VisualBasic/Portable/Binding/Binder_Expressions.vb#L4897-L4901

Reply:
Given that code, it looks like it's behaving as expected. Maybe @AlekseyTs knows the reason why it binds as it does?

Reply:
This might be an artifact of how native compiler behaved. It looks like changing this behavior affected only IOperation tests in the expected way. Feel free to open an issue for Roslyn.